cmd/kubeadm/app/cmd/upgrade/plan_test.go

```diff
@@ -688,8 +689,8 @@ func TestPrintUpgradePlanStructured(t *testing.T) {
 			expected: `apiVersion: output.kubeadm.k8s.io/v1alpha3
 availableUpgrades:
 - components:
-  - currentVersion: 1 x v1.8.1
-    name: kubelet
+  - currentVersion: v1.8.1
+    name: kubelet (node1)
```

Comment:
Should we add a new field to the ComponentUpgradePlan structure, i.e. NodeName? The new field should be filled with the name of the node that the component is running on only if the component is kubelet. Otherwise, the field should be empty.

Reply:
that does seem better, but it would be better to try to get the "node" for all components.

Reply:
Do you mean that this PR should get component manifests (3 x kube-apiserver、3 x controller、3 x scheduler) from apiserver, not from local directory?

The addons components are deployed by daemonset or deployment, so we shouldn't applied nodeName to their componentupgradeplan.

Reply:
> The addons components are deployed by daemonset or deployment, so we shouldn't apply nodeName to their componentupgradeplan.

yes, nodename can be empty for them (also the json field can have `omitempty`)

> Do you mean that this PR should get component pod manifests (3 x kube-apiserver、3 x controller、3 x scheduler) from apiserver, not from local directory?

with a combination of fieldselector and nodeselector we can get all control-plane TIER mirror pods and get their versions and node names

https://github.com/search?q=repo%3Akubernetes%2Fkubernetes%20ControlPlaneTier&type=code